File: src/formfiller.py
"""
FormMaster - Automated form filling for university applications
"""
import os
import argparse
import subprocess
import socket
import time
import sys
from threading import Lock
from pathlib import Path

from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import WebDriverException, NoSuchWindowException
from webdriver_manager.chrome import ChromeDriverManager
from selenium.webdriver import Firefox, FirefoxOptions
from pynput.mouse import Listener as MouseListener

from files.docxv1 import load
from forms.mod1 import mod1
from utils.logger import get_logger

# Global variables
lock = Lock()
IS_WINDOWS = os.name == 'nt'
driver = None
module = None
run_mode = 0
logger = get_logger('formfiller')

def setup_browser():
    """Initialize and setup the browser"""
    global driver
    
    if IS_WINDOWS:
        # Try to connect to an existing Chrome instance or start a new one
        if not connect_to_chrome_instance():
            start_chrome_instance()
        
        # Setup Chrome with debugging options
        chrome_options = Options()
        chrome_options.add_experimental_option("debuggerAddress", "127.0.0.1:9222")
        
        # Try to use local ChromeDriver if available
        user_profile = os.environ.get('USERPROFILE', '')
        local_driver_path = os.path.join(user_profile, '.formmaster', 'chromedriver.exe')
        
        if os.path.exists(local_driver_path):
            logger.info(f"Using local ChromeDriver at {local_driver_path}")
            service = Service(executable_path=local_driver_path)
            driver = webdriver.Chrome(service=service, options=chrome_options)
        else:
            logger.info("Using ChromeDriverManager")
            driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()), 
                                     options=chrome_options)
    else:
        # Firefox setup for non-Windows platforms
        options = FirefoxOptions()
        options.set_preference("network.protocol-handler.external-default", False)
        options.set_preference("network.protocol-handler.expose-all", True)
        options.set_preference("network.protocol-handler.warn-external-default", False)
        driver = Firefox(options=options)

def connect_to_chrome_instance():
    """Try to connect to an existing Chrome instance on port 9222"""
    server_address = ('127.0.0.1', 9222)
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
        sock.connect(server_address)
        logger.info("Connected to existing Chrome instance")
        return True
    except:
        return False
    finally:
        sock.close()

def start_chrome_instance():
    """Start a new Chrome instance with remote debugging enabled"""
    logger.info('Starting new Chrome browser instance...')
    
    # Find Chrome executable
    chrome_paths = []
    
    if IS_WINDOWS:
        for basedir in ['ProgramFiles', 'ProgramFiles(x86)', 'LocalAppData']:
            if basedir in os.environ:
                path = f"{os.environ[basedir]}\\Google\\Chrome\\Application\\chrome.exe"
                chrome_paths.append(path)
    else:
        chrome_paths.append('/opt/google/chrome/chrome')
    
    # Try each path until we find a valid Chrome executable
    for chrome in chrome_paths:
        if os.path.isfile(chrome):
            profile_dir = os.path.join(
                os.environ['LocalAppData'] if IS_WINDOWS else os.environ['HOME'],
                'selenium', 'ChromeProfile'
            )
            
            # Ensure profile directory exists
            Path(profile_dir).mkdir(parents=True, exist_ok=True)
            
            cmd = [chrome, '--remote-debugging-port=9222', f'--user-data-dir={profile_dir}']
            logger.info(f'Starting browser: {cmd}')
            subprocess.Popen(cmd)
            
            # Give browser time to start
            time.sleep(2)
            return
    
    logger.error("Could not find Chrome browser executable")
    sys.exit(1)

def is_browser_alive():
    """Check if browser is still open and responsive"""
    if not driver:
        return False
        
    try:
        # Attempt to access a property that requires the browser to be open
        driver.current_window_handle
        return True
    except (NoSuchWindowException, WebDriverException):
        try:
            driver.window_handles
            if len(driver.window_handles) > 0:
                driver.switch_to.window(driver.window_handles[-1])
                logger.info('Switching to latest window/tab')
                return True
            else:
                logger.info("Browser was closed by user or crashed")
                return False
        except:
            logger.info("Browser was closed by user or crashed")
            return False

def on_click(x, y, button, pressed):
    """Handle mouse click events"""
    global driver, module
    
    # Only process release events
    if pressed:
        return
    
    # Check if browser is still alive
    if not is_browser_alive():
        logger.warning("Browser closed, exiting application")
        os._exit(0)  # Force exit since we're in a threaded context
    
    try:
        with lock:
            if button.name == 'middle':
                # Middle click triggers module run
                module.run()
                return
            elif button.name == 'left':
                # Left click ensures we're on the most recent window/tab
                try:
                    wait = WebDriverWait(driver, 3)
                    wait.until(EC.presence_of_element_located((By.TAG_NAME, "body")))
                    
                    if driver.window_handles[-1] != driver.current_window_handle:
                        logger.info('Switching to latest window/tab')
                        driver.switch_to.window(driver.window_handles[-1])
                except Exception as e:
                    logger.debug(f"Window switching error (non-critical): {e}")
                return
    except Exception as e:
        logger.error(f"Error in on_click: {str(e)}")
        
        # Try to recover by switching to last window handle
        try:
            if len(driver.window_handles) > 0:
                driver.switch_to.window(driver.window_handles[-1])
            else:
                logger.critical("No browser windows available, exiting")
                os._exit(1)
        except:
            logger.critical("Fatal error in mouse handler, exiting")
            os._exit(1)

def run(data_dir, mode=0):
    """Main execution function"""
    global module, driver, run_mode
    
    run_mode = mode
    
    # Set up the browser
    setup_browser()
    
    # Load student data if in normal mode
    students = []
    if not run_mode:
        logger.info(f"Loading student data from {data_dir}")
        students = load(data_dir)
        logger.info(f"Loaded {len(students)} student records")

    # Initialize the appropriate module
    logger.info("Initializing Sydney University module")
    module = mod1(driver, students, run_mode)

    # Login and initiate the session
    main_application_handle = module.login_session()

    try:
        # Start mouse listener
        logger.info("Starting mouse listener")
        mouse_listener = MouseListener(on_click=on_click)
        mouse_listener.start()

        # Main event loop
        logger.info("Running main loop - waiting for events")
        while is_browser_alive():
            time.sleep(5)
            
        logger.info("Browser closed, exiting application")
        
    except KeyboardInterrupt:
        logger.info("Application interrupted by user")
    except Exception as e:
        logger.exception("Unhandled exception in main loop")
    finally:
        # Clean up
        logger.info("Stopping mouse listener")
        if 'mouse_listener' in locals():
            mouse_listener.stop()

def parse_arguments():
    """Parse command line arguments"""
    default_data_dir = 'C:\\work\\data\\13. 懿心ONE Bonnie' if IS_WINDOWS else '/home/hmei/data/13. 懿心ONE Bonnie'
    
    parser = argparse.ArgumentParser(description='FormMaster - Automate form filling for university applications.')
    
    parser.add_argument('--dir', type=str, 
                      default=default_data_dir,
                      help='Directory containing student data')
    
    parser.add_argument('--mode', type=int, default=0,
                      help='Operation mode (0 for normal operation)')
    
    return parser.parse_args()

if __name__ == '__main__':
    args = parse_arguments()
    run(data_dir=args.dir, mode=args.mode)

File: js-src/formfiller.js
/**
 * FormMaster - Automated form filling for university applications
 */
const os = require('os');
const path = require('path');
const fs = require('fs');
const { execSync } = require('child_process');
const net = require('net');
const { Builder, By, until } = require('selenium-webdriver');
const chrome = require('selenium-webdriver/chrome');
const firefox = require('selenium-webdriver/firefox');
const { pynput } = require('robotjs'); // Equivalent for pynput
const { loadData } = require('./files/docxv1');
const Mod1 = require('./forms/mod1');
const { getLogger } = require('./utils/logger');

// Global variables
let lock = new Mutex();
const IS_WINDOWS = os.platform() === 'win32';
let driver = null;
let module = null;
let runMode = 0;
const logger = getLogger('formfiller');

/**
 * Mutex implementation for locking
 */
function Mutex() {
  let isLocked = false;
  this.acquire = () => {
    if (isLocked) return false;
    isLocked = true;
    return true;
  };
  this.release = () => {
    isLocked = false;
  };
}

/**
 * Initialize and setup the browser
 */
async function setupBrowser() {
  if (IS_WINDOWS) {
    // Try to connect to an existing Chrome instance or start a new one
    if (!await connectToChromeInstance()) {
      await startChromeInstance();
    }
    
    // Setup Chrome with debugging options
    const chromeOptions = new chrome.Options();
    chromeOptions.addArguments('--remote-debugging-port=9222');
    
    // Try to use local ChromeDriver if available
    const userProfile = process.env.USERPROFILE || '';
    const localDriverPath = path.join(userProfile, '.formmaster', 'chromedriver.exe');
    
    if (fs.existsSync(localDriverPath)) {
      logger.info(`Using local ChromeDriver at ${localDriverPath}`);
      const service = new chrome.ServiceBuilder(localDriverPath).build();
      driver = await new Builder()
        .forBrowser('chrome')
        .setChromeOptions(chromeOptions)
        .setChromeService(service)
        .build();
    } else {
      logger.info("Using default ChromeDriver");
      driver = await new Builder()
        .forBrowser('chrome')
        .setChromeOptions(chromeOptions)
        .build();
    }
  } else {
    // Firefox setup for non-Windows platforms
    const options = new firefox.Options();
    options.setPreference("network.protocol-handler.external-default", false);
    options.setPreference("network.protocol-handler.expose-all", true);
    options.setPreference("network.protocol-handler.warn-external-default", false);
    
    driver = await new Builder()
      .forBrowser('firefox')
      .setFirefoxOptions(options)
      .build();
  }
  
  return driver;
}

/**
 * Try to connect to an existing Chrome instance on port 9222
 */
async function connectToChromeInstance() {
  return new Promise((resolve) => {
    const client = net.connect({port: 9222, host: '127.0.0.1'}, () => {
      logger.info("Connected to existing Chrome instance");
      client.end();
      resolve(true);
    });
    
    client.on('error', () => {
      resolve(false);
    });
  });
}

/**
 * Start a new Chrome instance with remote debugging enabled
 */
async function startChromeInstance() {
  logger.info('Starting new Chrome browser instance...');
  
  // Find Chrome executable
  const chromePaths = [];
  
  if (IS_WINDOWS) {
    for (const basedir of ['ProgramFiles', 'ProgramFiles(x86)', 'LocalAppData']) {
      if (process.env[basedir]) {
        const path = `${process.env[basedir]}\\Google\\Chrome\\Application\\chrome.exe`;
        chromePaths.push(path);
      }
    }
  } else {
    chromePaths.push('/opt/google/chrome/chrome');
  }
  
  // Try each path until we find a valid Chrome executable
  for (const chrome of chromePaths) {
    if (fs.existsSync(chrome)) {
      const profileDir = path.join(
        IS_WINDOWS ? process.env.LocalAppData : process.env.HOME,
        'selenium', 'ChromeProfile'
      );
      
      // Ensure profile directory exists
      if (!fs.existsSync(profileDir)) {
        fs.mkdirSync(profileDir, { recursive: true });
      }
      
      const cmd = `"${chrome}" --remote-debugging-port=9222 --user-data-dir="${profileDir}"`;
      logger.info(`Starting browser: ${cmd}`);
      
      try {
        // Use spawn to start the process without waiting
        require('child_process').spawn(chrome, [
          '--remote-debugging-port=9222',
          `--user-data-dir=${profileDir}`
        ], {
          detached: true,
          stdio: 'ignore'
        }).unref();
        
        // Give browser time to start
        await new Promise(resolve => setTimeout(resolve, 2000));
        return;
      } catch (e) {
        logger.error(`Error starting Chrome: ${e.message}`);
      }
    }
  }
  
  logger.error("Could not find Chrome browser executable");
  process.exit(1);
}

/**
 * Check if browser is still open and responsive
 */
async function isBrowserAlive() {
  if (!driver) {
    return false;
  }
  
  try {
    // Attempt to access a property that requires the browser to be open
    await driver.getWindowHandle();
    return true;
  } catch (e) {
    try {
      const handles = await driver.getAllWindowHandles();
      if (handles.length > 0) {
        await driver.switchTo().window(handles[handles.length - 1]);
        logger.info('Switching to latest window/tab');
        return true;
      } else {
        logger.info("Browser was closed by user or crashed");
        return false;
      }
    } catch {
      logger.info("Browser was closed by user or crashed");
      return false;
    }
  }
}

/**
 * Handle mouse click events
 */
async function onClick(button, pressed) {
  // Only process release events
  if (pressed) return;
  
  // Check if browser is still alive
  if (!await isBrowserAlive()) {
    logger.warning("Browser closed, exiting application");
    process.exit(0);
  }
  
  try {
    if (lock.acquire()) {
      try {
        if (button === 'middle') {
          // Middle click triggers module run
          await module.run();
        } else if (button === 'left') {
          // Left click ensures we're on the most recent window/tab
          try {
            await driver.wait(until.elementLocated(By.tagName("body")), 3000);
            
            const currentHandle = await driver.getWindowHandle();
            const handles = await driver.getAllWindowHandles();
            
            if (handles[handles.length - 1] !== currentHandle) {
              logger.info('Switching to latest window/tab');
              await driver.switchTo().window(handles[handles.length - 1]);
            }
          } catch (e) {
            logger.debug(`Window switching error (non-critical): ${e.message}`);
          }
        }
      } finally {
        lock.release();
      }
    }
  } catch (e) {
    logger.error(`Error in onClick: ${e.message}`);
    
    // Try to recover by switching to last window handle
    try {
      const handles = await driver.getAllWindowHandles();
      if (handles.length > 0) {
        await driver.switchTo().window(handles[handles.length - 1]);
      } else {
        logger.critical("No browser windows available, exiting");
        process.exit(1);
      }
    } catch {
      logger.critical("Fatal error in mouse handler, exiting");
      process.exit(1);
    }
  }
}

/**
 * Main execution function
 */
async function run(dataDir, mode = 0) {
  runMode = mode;
  
  // Set up the browser
  driver = await setupBrowser();
  
  // Load student data if in normal mode
  let students = [];
  if (!runMode) {
    logger.info(`Loading student data from ${dataDir}`);
    students = await loadData(dataDir);
    logger.info(`Loaded ${students.length} student records`);
  }

  // Initialize the appropriate module
  logger.info("Initializing Sydney University module");
  module = new Mod1(driver, students, runMode);

  // Login and initiate the session
  const mainApplicationHandle = await module.loginSession();

  try {
    // Start mouse listener with robotjs
    logger.info("Starting mouse listener");
    
    // Main event loop
    logger.info("Running main loop - waiting for events");
    while (await isBrowserAlive()) {
      await new Promise(resolve => setTimeout(resolve, 5000));
    }
    
    logger.info("Browser closed, exiting application");
    
  } catch (e) {
    if (e.code === 'SIGINT') {
      logger.info("Application interrupted by user");
    } else {
      logger.error(`Unhandled exception in main loop: ${e.message}`);
      logger.error(e.stack);
    }
  } finally {
    // Clean up
    logger.info("Cleaning up resources");
    
    // Close the browser if it's still open
    if (driver) {
      try {
        await driver.quit();
      } catch (e) {
        logger.debug(`Error closing browser: ${e.message}`);
      }
    }
  }
}

/**
 * Parse command line arguments
 */
function parseArguments() {
  const defaultDataDir = IS_WINDOWS ? 
    'C:\\work\\data\\13. 懿心ONE Bonnie' : 
    '/home/hmei/data/13. 懿心ONE Bonnie';
  
  // Simple argument parser
  const args = process.argv.slice(2);
  let dataDir = defaultDataDir;
  let mode = 0;
  
  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--dir' && i + 1 < args.length) {
      dataDir = args[i + 1];
      i++;
    } else if (args[i] === '--mode' && i + 1 < args.length) {
      mode = parseInt(args[i + 1], 10);
      i++;
    }
  }
  
  return { dataDir, mode };
}

// Export the run function
module.exports = { run };

// Run if called directly
if (require.main === module) {
  const args = parseArguments();
  run(args.dataDir, args.mode).catch(err => {
    console.error(`Fatal error: ${err.message}`);
    console.error(err.stack);
    process.exit(1);
  });
}
